compiler/resolution/functionResolution.cpp

```diff
@@ -11255,6 +11255,82 @@ static void checkNoVoidFields()
   }
 }
 
+/************************************* | **************************************
+*                                                                             *
+*  Find specially named methods such as 'hash' that are placed on types that  *
+*  don't have the corresponding interface. This is executed here because      *
+*  resolution cleans up unused functions and interface statements, but we want*
+*  to warn even if a function isn't used.                                     *
+*                                                                             *
+************************************** | *************************************/
+
+struct SpeciallyNamedMethodInfo {
+  std::map<std::string, FnSymbol*> speciallyNamedMethods;
+};
+
+using SpeciallyNamedMethodKey = std::pair<AggregateType*, InterfaceSymbol*>;
+using SpecialMethodMap = std::map<SpeciallyNamedMethodKey, SpeciallyNamedMethodInfo>;
+
+static void checkSpeciallyNamedMethods() {
+  static const std::unordered_map<std::string, InterfaceSymbol*> reservedNames = {
+    { "hash", gHashable },
```

Comment:
You might want to use an `astr` on "hash" due to considerations about string comparisons, assuming there isn't one already.

I unfortunately don't remember the full details of why that's necessary, but it's something that's bit me in the past so I figured I should bring it up

Reply:
I used `std::string` to avoid the need for `astr` (`astr` does interning, so allows `==` on `const char*`; `==` works as we'd expect for `std::string`). I wasn't sure if `name` on `FnSymbol` is interned or not; if it wasn't, it would need to be interned too, which I think would be slower than just constructing a string.

I'll check if `FnSymbols` intern their strings; if they do, using `astr` here would be better.

Reply:
They do intern their strings, so using `astr` on `hash` ad `const char*` instead of `std::string` should work here.